Makefile

```diff
@@ -139,30 +139,41 @@ frontend-lint: node_modules/.uptodate
 bddtests: python
 	@tox -qe bddtests
 
+# Find the prerequisites ('-r foo.txt' lines) in a requirements/*.in file.
+#
+# This is a sed command of the form:
+#
+#     sed -n 's/<regex>/<replacement>/p'
+#
+# * The -n tells sed not to print the non-matching lines
+# * The /p tells sed to print the matching lines
+# * The regex part finds '-r foo.txt' or '-c bar.txt' lines
+# * The replacement part prints 'requirements/foo.txt' with the -r or -c,
+#   whitespace and comments removed
+# * The tr at the very end joins the output into one line
+find_prerequisites = $(shell sed -n 's/^[[:blank:]]*\(-r\|-c\)[[:blank:]]*\([[:alnum:]\/-_\.]\+\).*$$/requirements\/\2/p' "requirements/$*.in" | tr '\n' ' ')
```

Comment:
Given this input file:

```
# These normal package names should not be printed:
pip-tools
toml # Needed for pydocstyle to support pyproject.toml.
pylint
pydocstyle
pycodestyle

# These imports should be printed with "requirements/" prepended and their
# -r's, -c's, whitespace and comments stripped:
-r tests.txt
    -r  functests.txt
-rbddtests.txt#Trailing comment
 -cconstraints.txt # Trailing comment

# These lines that look like imports but aren't should not be printed:
# -r bar.txt
"-rbar.txt"

# These -r's have no filename after them, they shouldn't be printed.
-r
-r# Trailing comment
```

The command outputs:

    requirements/tests.txt requirements/functests.txt requirements/bddtests.txt requirements/constraints.txt

Reply:
I think this would be better as a separate script which would be:

* More familiar to Python developers
* Require much less explanation
* Be testable if we wanted to
* About the same size overall when comments are included

This could still be pretty small. You could make this script smaller, but I'm not sure it's worth it.

```python
#!/usr/bin/env python
"""Find all items in a requirements file marked with `-r` or `-c`."""

from os.path import dirname, join
from pathlib import Path
import re
import sys


def main(file_name):
    requirement_files = re.findall(
        r"^\s*-[cr]\s*([^\s#]+)", Path(file_name).read_text(), re.MULTILINE
    )

    print(" ".join(join(dirname(file_name), i) for i in requirement_files))


if __name__ == "__main__":
    main(file_name=sys.argv[1])
```

I'm not sure how this would integrate with the make variable interpolation. Part of this would make the boundary clearer.

Reply:
> I think this would be better as a separate script which would be:
> 
> * More familiar to Python developers

I think that's very reasonable.

Not tested much but I think this is how you'd integrate a separate `bin/find_prerequisites` script into the `Makefile`:

```make
.SECONDEXPANSION:
requirements/%.txt: requirements/%.in $$(shell bin/find_prerequisites requirements/$$*.in)
	@touch -a $(subst requirements.txt,dev.txt,$@)
	@tox -qe $(subst requirements,dev,$(basename $(notdir $@))) --run-command 'pip-compile --allow-unsafe --quiet $(args) $<'
```

I wonder what it'd be like if it were a separate script file but was a bash script still using sed and tr? But I think Python is very reasonable as well.